packages/core/src/hydration/error_handling.ts

```diff
@@ -6,39 +6,366 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {TNode} from '../render3/interfaces/node';
+import {RuntimeError, RuntimeErrorCode} from '../errors';
+import {getDeclarationComponentDef} from '../render3/instructions/element_validation';
+import {TNode, TNodeType} from '../render3/interfaces/node';
 import {RNode} from '../render3/interfaces/renderer_dom';
-import {LView} from '../render3/interfaces/view';
+import {LView, TVIEW} from '../render3/interfaces/view';
+import {getParentRElement} from '../render3/node_manipulation';
+
+import {NodeNavigationStep} from './interfaces';
+
+
+const AT_THIS_LOCATION = '<-- AT THIS LOCATION';
+
+function getFriendlyStringFromTNodeType(key: number): string {
+  switch (key) {
+    case TNodeType.Container:
+      return 'view container';
+    case TNodeType.Element:
+      return 'element';
+    case TNodeType.ElementContainer:
+      return 'ng-container';
+    case TNodeType.Icu:
+      return 'icu';
+    case TNodeType.Placeholder:
+      return 'i18n';
+    case TNodeType.Projection:
+      return 'projection';
+    case TNodeType.Text:
+      return 'text';
+    default:
+      return '';
+  }
+}
 
 /**
- * Verifies whether a given node matches an expected criteria,
- * based on internal data structure state.
+ * Validates that provided nodes match during the hydration process.
  */
 export function validateMatchingNode(
-    node: RNode, nodeType: number, tagName: string|null, lView: LView, tNode: TNode): void {
-  validateNodeExists(node);
-  if ((node as Node).nodeType !== nodeType ||
-      (node as Node).nodeType === Node.ELEMENT_NODE &&
-          (node as HTMLElement).tagName.toLowerCase() !== tagName?.toLowerCase()) {
-    // TODO: improve error message and use RuntimeError instead.
-    throw new Error(`Unexpected node found during hydration.`);
+    node: RNode, nodeType: number, tagName: string|null, lView: LView, tNode: TNode,
+    isViewContainerAnchor = false): void {
+  if (!node ||
+      ((node as Node).nodeType !== nodeType ||
+       ((node as Node).nodeType === Node.ELEMENT_NODE &&
+        (node as HTMLElement).tagName.toLowerCase() !== tagName?.toLowerCase()))) {
```

Comment:
Once the `validateNodeExists` is restored above, we can extract the `nodeType` outside of the `if` (the node would always exists):
```suggestion
      const nodeType = (node as Node).nodeType;
      if (nodeType !== nodeType || (nodeType === Node.ELEMENT_NODE &&
        (node as HTMLElement).tagName.toLowerCase() !== tagName?.toLowerCase()))) {
```